Align `aws-http4s` module on the constructs now provided by `http4s-kernel`

Baccata opened the pull request:
This PRs aims at standardising the AWS support on http4s. Trying to support several clients and idioms would induce a perfomance burden that I'm not willing to carry, and hurt our ability to support streaming operations that some AWS services expose (in particular S3). 

* Removes the `aws` module. Its responsibilities move to the `aws-http4s` module.
* Overhauls most of the logic in order to benefit from the granular constructs introduced `http4s-kernel`
* In particular, signature now occurs via a `Client => Client` middleware. This drastically improves maintainability of the code at the cost of a little bit of indirection. This approach will make it easier to support other means of signature (in particular for streaming operations). 
* Changes the UX for instantiating clients. In particular, the instantiation is now biased towards monofunctors, as most of the interaction with AWS is unary. But also, implicit extensions are removed in favour of an `AwsClient(_, _)` call, which makes it easier to rely on auto-completion in editors. 

TODO : 

- [x] Add unit-tests for the signature middleware
- [ ] Polish UX around client instantiation 
- [ ] Add Scaladocs to the `AwsClient` construct, the `AwsRegion` construct and the`AwsEnvironment` construct, which

scala-steward committed:
Update coursier to 2.1.0-RC6

yisraelU committed:
aws tests

yisraelU committed:
aws tests

yisraelU committed:
client uses withEmptyBody instead of withEntity when there is no body

yisraelU committed:
Merge branch 'dynamic-complianceTests' into aws-tests

yisraelU committed:
minor refactor

yisraelU committed:
handle empty body field in protocol tests

yisraelU committed:
handle empty fields in testcase

yisraelU committed:
Merge branch 'dynamic-complianceTests' into aws-tests

yisraelU committed:
Merge branch 'dynamic-complianceTests' into aws-tests

# Conflicts:
#	modules/compliance-tests/src/smithy4s/compliancetests/internals/Assertions.scala
#	modules/compliance-tests/src/smithy4s/compliancetests/internals/ClientHttpComplianceTestCase.scala

scala-steward committed:
Update jsoniter-scala-core to 2.21.2

scala-steward committed:
Update swagger-ui-dist to 4.16.1

scala-steward committed:
Update smithy-build, smithy-model to 1.28.0

yisraelU committed:
Merge branch 'dynamic-complianceTests' into aws-tests

yisraelU committed:
changed timestamp handling in Jcodec to allow Aws Codec to select default

yisraelU committed:
fmt

dependabot[bot] committed:
Bump @docusaurus/preset-classic from 2.3.0 to 2.3.1 in /modules/website

Bumps [@docusaurus/preset-classic](https://github.com/facebook/docusaurus/tree/HEAD/packages/docusaurus-preset-classic) from 2.3.0 to 2.3.1.
- [Release notes](https://github.com/facebook/docusaurus/releases)
- [Changelog](https://github.com/facebook/docusaurus/blob/main/CHANGELOG.md)
- [Commits](https://github.com/facebook/docusaurus/commits/v2.3.1/packages/docusaurus-preset-classic)

---
updated-dependencies:
- dependency-name: "@docusaurus/preset-classic"
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump @docusaurus/core from 2.3.0 to 2.3.1 in /modules/website

Bumps [@docusaurus/core](https://github.com/facebook/docusaurus/tree/HEAD/packages/docusaurus) from 2.3.0 to 2.3.1.
- [Release notes](https://github.com/facebook/docusaurus/releases)
- [Changelog](https://github.com/facebook/docusaurus/blob/main/CHANGELOG.md)
- [Commits](https://github.com/facebook/docusaurus/commits/v2.3.1/packages/docusaurus)

---
updated-dependencies:
- dependency-name: "@docusaurus/core"
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

yisraelU committed:
Merge branch 'dynamic-complianceTests' into aws-tests

yisraelU committed:
fix float and double eq for Nan

yisraelU committed:
add filter for NullOeration

yisraelU committed:
comment

plokhotnyuk committed:
Add parsing of time zones for `DATE_TIME` format of `Timestamp`

plokhotnyuk committed:
Fix #750

Baccata committed:
Fix tests in Scala 2.12

yisraelU committed:
update alt

yisraelU committed:
headers added

yisraelU committed:
fmt

yisraelU committed:
update docs

yisraelU committed:
add host as input to test

kubukoz committed:
Fix description of date formats

kubukoz committed:
Fix description of date formats

Baccata committed:
WIP

kubukoz committed:
fix hour format description

kubukoz committed:
fix hour format description

Baccata committed:
Merge remote-tracking branch 'origin/series/0.18' into aws-alignement-on-http4s-kernel

Baccata committed:
Remove aws module

Baccata committed:
Cleanup + fix docs

Baccata committed:
Re-introduce the fixes to the aws-specific json codecs

Baccata commented on build.sbt:
```diff
@@ -301,37 +300,6 @@ lazy val `aws-kernel` = projectMatrix
-/**
- * cats-effect specific abstractions of AWS protocol interpreters and constructs
- */
-lazy val aws = projectMatrix
```
removed as supporting non-http4s clients in not in the scope of the project anymore.

Baccata commented on modules/aws-http4s/src/smithy4s/aws/AwsClient.scala:
```diff
@@ -0,0 +1,101 @@
+    private def interpreter[F[_]: Async](
+        awsEnv: AwsEnvironment[F]
+    ): service.FunctorInterpreter[F] = {
+      val clientCodecs: UnaryClientCodecs[F] = awsProtocol match {
```
this is where the money-shot is, which entirely justifies everything that 0.18 is about : the difference in protocols is entirely driven by the codecs. The rest of the logic is entirely decoupled from it, be that signature / routing / etc / etc.

Baccata committed:
Fix AwsJsonCodecs hint mask

Baccata commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsJsonCodecs.scala:
```diff
@@ -0,0 +1,46 @@
+/**
+ * An client codec for the AWS_JSON_1.0/AWS_JSON_1.1 protocol
+ */
+private[aws] object AwsJsonCodecs {
```
Putting aside the problem of streaming operations, this small construct centralises the protocol-specific details.

Baccata commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsJsonCodecs.scala:
```diff
@@ -0,0 +1,46 @@
+    aws.protocols.AwsJson1_0.protocol.hintMask ++
+      aws.protocols.AwsJson1_0.protocol.hintMask ++ HintMask(IntEnum)
+
+  def make[F[_]: Async]: UnaryClientCodecs[F] = {
```
I think there's still a little bit of polishing to do there. I think the `UnaryClientCodecs` need to be revised to become a mere holder of the compiled codecs that are relevant to the client. Something like : 

```scala
trait UnaryClientCodecs[F[_], I, E, O, SI, SO] {
  def inputEncoder : RequestEncoder[F, I]
  def outputDecoder : ResponseDecoder[F, O]
  def errorDecoder : ResponseDecoder[F, Throwable]
} 

object UnaryClientCodecs {
  type For[F[_]] {
     type toKind5[I, E, O, SI, SO] 
  } 
} 
```

And the `make` here should be something like : 

```scala
def make[F[_] : Async] : PolyFunction5[Endpoint.Schemas, UnaryClientCodecs.For[F]#toKind5] 
```

The reason is that the `Schemas` might not be sufficient to produce the encoders : some information which lives at the Endpoint level (like the `@endpoint` trait) can be necessary to produce it.

kubukoz replied:
> The reason is that the Schemas might not be sufficient to produce the encoders : some information which lives at the Endpoint level (like the @endpoint trait) can be necessary to produce it.

Why not accept Endpoint directly, then?

Baccata committed:
Headers

Baccata committed:
Extract signing function from signing client middleware

yisraelU committed:
merge aws-alignement-on-http4s-kernel

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsJsonCodecs.scala:
```diff
@@ -0,0 +1,46 @@
+private[aws] object AwsJsonCodecs {
+
+  private val hintMask =
+    aws.protocols.AwsJson1_0.protocol.hintMask ++
+      aws.protocols.AwsJson1_0.protocol.hintMask ++ HintMask(IntEnum)
```
**issue:** `AwsJson1_0` shows up twice, that seems like a mistake.

Did you want to use 1_0 and 1_1? If so, should we really have that in a single mask?

kubukoz replied:
I guess they're supposed to be the same, let's just fix this to use both then

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsSignatureClient.scala:
```diff
@@ -0,0 +1,133 @@
+      kSigning
+    }
+
+    // scalafmt: { align.preset = most, danglingParentheses.preset = false, maxColumn = 240, align.tokens = [{code = ":"}]}
```
**praise:** TIL

kubukoz commented on build.sbt:
```diff
@@ -342,25 +310,35 @@ lazy val aws = projectMatrix
 lazy val `aws-http4s` = projectMatrix
   .in(file("modules/aws-http4s"))
-  .dependsOn(aws, complianceTests % "test->compile", dynamic % "test->compile")
+  .dependsOn(`aws-kernel`, `http4s-kernel`, json, xml, complianceTests % "test->compile", dynamic % "test->compile")
```
**issue:** I don't think we actually use either `dynamic` or `complianceTests` here.

kubukoz commented on build.sbt:
```diff
@@ -342,25 +310,35 @@ lazy val aws = projectMatrix
+      "smithy4s.example.aws"
+    ),
+    Test / sourceGenerators := Seq(genSmithyScala(Test).taskValue),
+    scalacOptions ++= Seq(
+      "-Wconf:msg=class AwsQuery in package (aws\\.)?protocols is deprecated:silent"
+    )
```
**praise:** nice

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/AwsClient.scala:
```diff
@@ -0,0 +1,101 @@
+        case AwsProtocol.AWS_JSON_1_1(_) =>
+          // TODO "application/x-amz-json-1.1",
+          AwsJsonCodecs.make[F]
+        case _ => ???
```
**issue:** let's `sys.error` this or throw something explicitly

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/AwsClient.scala:
**suggestion:** not strictly related, but can we have comments on `RequestEncoder` / `ResponseDecoder` pointing to `MessageEncoder` and `MessageDecoder` as a place where some built-in implementations can be found? It took me a sec and it might just make it more discoverable.

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/AwsClient.scala:
```diff
@@ -0,0 +1,101 @@
+          def apply[I, E, O, SI, SO](
+              endpoint: service.Endpoint[I, E, O, SI, SO]
+          ): I => F[O] =
+            new AwsUnaryEndpoint(
+              service.id,
+              service.hints,
+              awsService,
+              awsEnv,
+              endpoint,
+              clientCodecs
+            )
```
**suggestion:** named params

kubukoz committed:
Async -> Concurrent, remove unused build dependencies

kubukoz committed:
Fix some outstanding Asyncs

kubukoz commented:
Playground is now happy with this: https://github.com/kubukoz/smithy-playground/pull/255/files - look at how smol it is now with the Async change reverted :D

kubukoz committed:
Generate Schema.apply overloads as returning `Schema.StructSchema`

yisraelU committed:
update aws-http4s compliance tests

yisraelU committed:
content type fix for AwsSignatureClient

yisraelU committed:
header added

Baccata committed:
Merge pull request #920 from disneystreaming/fix-content-header

[Aws-Http4s] Fix content header for Aws Signature Client

Baccata committed:
Merge branch 'aws-alignement-on-http4s-kernel' of https://github.com/disneystreaming/smithy4s into aws-alignement-on-http4s-kernel

Baccata committed:
Account for TCP routing when testing how routers decode requests.

Avoid solely relying on intercepting IntendedShortCircuit exceptions
when testing routers, as this method doesn't account for routers that
exercise the TCP layer.

Instead, we still try to intercept that exception (which will work for
routers built on top of Http4s, or routers that do an in-memory conversion
to the Http4s models), but we now also try to intercept 5xx errors before
waiting on a promise. A Router exercises the TCP layer will be likely
to turn uncaught exceptions (like the one we raise) into a 500 http
response.

jgoday committed:
#913: Make generated case classes final

jgoday committed:
Update modules/example/src autogenerated files

Commit:
Fix fmt

Baccata committed:
Adds tests to verify our implementation of the AWS signature algorithm

yisraelU committed:
add jttpResonseCode support for int enum

yisraelU committed:
Merge branch 'aws-alignement-on-http4s-kernel' into aws-tests

yisraelU committed:
add back module dependencies for aws tests
fix incorrect ordering for header test result message

Baccata committed:
Fixed query params ordering in AWS signature

Also moved tests so that they are executed on Native and JS

Baccata committed:
Testing signature algorithm with access token

Baccata committed:
Add the aws sdk to the list of things Scala Steward should ignore

Baccata committed:
Use

scala-steward committed:
Update fs2-core, fs2-io to 3.6.1

scala-steward committed:
Update jsoniter-scala-core to 2.23.0

scala-steward committed:
Update fs2-data-xml to 1.7.0

scala-steward committed:
Update http4s-circe, http4s-client, ... to 0.23.18

yisraelU committed:
add test for HttpResponseCodeSchemaVisitor

Baccata committed:
awssdk should be a test-only dep

Baccata committed:
Merge pull request #927 from disneystreaming/aws-signature-tests

Add AWS signature tests

kubukoz committed:
Update modules/core/test/src/smithy4s/http/internals/HttpResponseCodeSchemaVisitorSpec.scala

yisraelU committed:
add SmithyNodeDecoderSChemaVisitor

yisraelU committed:
update tests

yisraelU committed:
update tests

scala-steward committed:
Update swagger-ui-dist to 4.18.3

yisraelU committed:
updated to use ignored exception

scala-steward committed:
Update cats-effect to 3.4.10

Baccata committed:
Fix a few bugs, improving compliance in the process

Baccata committed:
Empty body is assumed to be empty json object in AWS

Baccata committed:
Filtering-out irrelevant tests

Baccata committed:
Merge remote-tracking branch 'origin/aws-alignement-on-http4s-kernel' into aws-tests

Baccata committed:
Sync branch, move compliance-tests in platform-agnostic folder

Baccata committed:
Simplifiy the canonical Node decoding override

scala-steward committed:
Update sbt-sonatype to 3.9.19

yisraelU committed:
update alloy dep

Baccata committed:
Bumped Alloy + filtered out compression-related tests

Baccata committed:
Regenerated examples

Baccata committed:
Remove dependency from docs to aws-http4s/Test

Baccata committed:
headers

Baccata committed:
Merge pull request #934 from disneystreaming/aws-tests

 Aws Protocol Compliance tests

Baccata commented:
I'm making the BDFL decision to merge this as is into 0.18 in order to unlock ongoing parallel work, and I'll address comments later
